Raise `ActiveRecord:: InverseOfAssociationNotFoundError` if invalid inverse_of is specified

alpaca-tc opened the pull request:
### Motivation / Background

This error occurs when the association specified in the `inverse_of` option does not exist on the associated class.
Previously this would be implicitly ignored, so the developer wouldn't know they tried to make an invalid association.

### Detail

```ruby
Post.belongs_to(:user, inverse_of: :comment) # Correct inverse_of is :post
user = User.create!

 # Before:
Post.new(user: user) #=> No error

 # After:
Post.new(user: user)
#=> ActiveRecord::InverseOfAssociationNotFoundError: Could not find the inverse association for user (:comment in User).
```

### Additional information

When checking the bi-directionality of inverse_of reflection, it is intuitive to
compare classes. However, the implementation compares `table_name` instead, because
class name comparisons are challenging with anonymous classes (such as those used in Rails testing).

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

alpaca-tc committed:
Raise `ActiveRecord::InverseOfAssociationNotFoundError` if invalid inverse_of is specified.

This error occurs when the association specified in the `inverse_of` option does not exist on the associated class.
Previously this would be implicitly ignored, so the developer wouldn't know they tried to make an invalid association.

```ruby
Post.belongs_to(:user, inverse_of: :comment) # Correct inverse_of is :post
user = User.create!

 # Before:
Post.new(user: user) #=> No error

 # After:
Post.new(user: user)
 #=> ActiveRecord::InverseOfAssociationNotFoundError: Could not find the inverse association for user (:comment in User).
```

When checking the bi-directionality of inverse_of reflection, it is intuitive to
compare classes. However, the implementation compares table_name instead, because
class name comparisons are challenging with anonymous classes (such as those used in Rails testing).

Co-authored-by: zzak <zzakscott@gmail.com>

yahonda commented:
Although English isn't my native language, I find the current error message like Inverse association human (:interests in Human) is mismatch table_name." unclear for me, I assume this statement is based on https://github.com/rails/rails/pull/41552 , where the "recursive" is an adjective, while "mismatch" is a noun.

alpaca-tc commented:
@yahonda I modified the class name of the exception has been changed from `InverseOfAssociationTableNameMismatchError` to `InverseOfAssociationNotBidirectionalError`.
This is because the old name is not essential, since it expresses implementation details.
In addition, the error message has been changed to point out that it is not bi-directional.

Also, the PR description has been modified to explain why `table_name` is used for bi-directional comparisons. Originally I said STI considered, but it was correct that it considered the anonymous class.

Please re-review this PR 😄 🙏

rafaelfranca commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -71,6 +71,19 @@ def initialize(reflection = nil)
+      if reflection
+        super("Inverse association #{reflection.name} (#{reflection.inverse_of.name.inspect} in #{reflection.class_name}) is not bi-directional.")
+      else
+        super("Inverse association is not bi-directional.")
```
Even I not sure what this error message means. It seems like it is a requirement for inverse_of to be by directional, but I don't think this is the problem here, is it? 

The problem is that the association on the other model doesn't exist, so we should be explicit about that.

Something like:

```
Inverse association doesn't not exist in the associated class
```

alpaca-tc replied:
@rafaelfranca  I have corrected the message you commented. 🙏 
Should I L81 have also been changed?

zzak commented on activerecord/test/cases/associations/inverse_associations_test.rb:
```diff
@@ -873,6 +873,25 @@ def test_recursive_inverse_on_recursive_model_has_many_inversing
+    # Correct inverse_of should be `:face`, but the incorrect configuration `:interests` was used.
+    face_klass.belongs_to(:human, inverse_of: :interests)
+
+    error = assert_raises(ActiveRecord::InverseOfAssociationNotBidirectionalError) do
```
You can use [assert_raises(match:)](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Testing/Assertions.html#method-i-assert_raises).

```
assert_raises(AR::IoANotFoundError, match: /Inverse association human (:association in Huma) does not exist./)
```

zzak commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,29 @@
+*   Raise `ActiveRecord::InverseOfAssociationNotBidirectionalError` if invalid inverse_of is specified
+
+    The association specified in `inverse_of` should be the name of the association
+    on the other model that points back to the current model.
+    If the user inadvertently specifies an invalid association name in `inverse_of`
+    that is not bi-directional, the `inverse_of` is implicitly ignored.
+    However, if it remains implicitly ignored, the user can't know that the inverse_of
+    is incorrect. Therefore, I have made a change to raise an exception instead of.
```
```suggestion
    This error occurs when the association specified in the `inverse_of` option does not exist on the associated class.
    Previously this would be implicitly ignored, so the developer wouldn't know they tried to make an invalid association.
```

zzak commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,29 @@
+    Post.new(user: user) #=> No error
+
+     # After:
+    Post.new(user: user) #=> ActiveRecord::InverseOfAssociationNotBidirectionalError: Inverse association user (:comment in User) is not bi-directional.
```
```suggestion
    Post.new(user: user)
    #=> ActiveRecord::InverseOfAssociationNotBidirectionalError: Inverse association user (:comment in User) is not bi-directional.
```

zzak commented on activerecord/CHANGELOG.md:
````diff
@@ -1,3 +1,29 @@
+     # After:
+    Post.new(user: user) #=> ActiveRecord::InverseOfAssociationNotBidirectionalError: Inverse association user (:comment in User) is not bi-directional.
+    ```
+
+    When checking the bi-directionality of inverse_of reflection, it is intuitive to
+    compare classes. However, the implementation compares table_name instead, because
+    class name comparisons are challenging with anonymous classes (such as those used in Rails testing).
````
```suggestion
```

This is implementation detail which should not go into release notes.

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -71,6 +71,20 @@ def initialize(reflection = nil)
     end
   end
 
+  class InverseOfAssociationNotBidirectionalError < ActiveRecordError # :nodoc:
```
I think there was already some discussion about the name, but I prefer `InverseOfAssociationNotFoundError`, as it's explicit that the "inverse of" doesn't exist.

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -71,6 +71,20 @@ def initialize(reflection = nil)
+      if reflection
+        super("Inverse association #{reflection.name} (#{reflection.inverse_of.name.inspect} in #{reflection.class_name}) is not bi-directional.")
+      else
+        super("Inverse association doesn't not exist in the associated class.")
```
```suggestion
        super("Inverse association doesn't exist in the associated class.")
```

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -71,6 +71,20 @@ def initialize(reflection = nil)
+      @reflection = reflection
+
+      if reflection
+        super("Inverse association #{reflection.name} (#{reflection.inverse_of.name.inspect} in #{reflection.class_name}) is not bi-directional.")
```
```suggestion
        super("Inverse association #{reflection.name} (#{reflection.inverse_of.name.inspect}) doesn't exist in #{reflection.class_name}.")
```

alpaca-tc commented:
Thank you for the review 🙏 

I have incorporated suggestions and modified the code to use `InverseOfAssociationNotFoundError`. The changes have been force-pushed. Please review the updates.

yahonda commented:
Would you address the conflict?

alpaca-tc commented:
@yahonda I rebased and force-pushed 🙏

alpaca-tc commented:
@rafaelfranca I’ve made the requested changes. When you have a moment, could you please review it again? 😄
